applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py

```diff
@@ -0,0 +1,576 @@
+    def __computeControlPointUpdate(self):
+        """adapted from https://msulaiman.org/onewebmedia/GradProj_2.pdf"""
+        if self.inner_iter == 1:
+            self.g_a, self.g_a_variables, self.relaxation_coefficients, self.correction_coefficients = self.__getActiveConstraints()
+
+        KM.Logger.PrintInfo("ShapeOpt", "Assemble vector of objective gradient.")
+        nabla_f = KM.Vector()
+        p = KM.Vector()
+        self.optimization_utilities.AssembleVector(self.design_surface, nabla_f, KSO.DF1DX_MAPPED)
+        f_norm = self.optimization_utilities.ComputeMaxNormOfNodalVariable(self.design_surface, KSO.DF1DX_MAPPED)
+
+        if abs(f_norm) > 1e-10:
+            nabla_f *= 1.0/f_norm
+
+        if len(self.g_a) == 0:
+            KM.Logger.PrintInfo("ShapeOpt", "No constraints active, use negative objective gradient as search direction.")
+            p = nabla_f * (-1.0)
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, p, KSO.SEARCH_DIRECTION)
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, p, KSO.PROJECTION)
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, [0.0]*len(p), KSO.CORRECTION)
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, p, KSO.CONTROL_POINT_UPDATE)
+            self.__LineSearch()
+            return
+
+        omega_r = KM.Matrix()
+        self.optimization_utilities.AssembleBufferMatrix(omega_r, self.relaxation_coefficients)
+        omega_c = KM.Vector()
+        self.optimization_utilities.AssembleBufferVector(omega_c, self.correction_coefficients)
+        KM.Logger.PrintInfo("ShapeOpt", "Assemble matrix of constraint gradient.")
+        N = KM.Matrix()
+        self.optimization_utilities.AssembleVectorstoMatrix(self.design_surface, N, self.g_a_variables)  # TODO check if gradients are 0.0! - in cpp
+
+        settings = KM.Parameters('{ "solver_type" : "LinearSolversApplication.dense_col_piv_householder_qr" }')
+        solver = dense_linear_solver_factory.ConstructSolver(settings)
+
+        c = KM.Vector()
+
+        KM.Logger.PrintInfo("ShapeOpt", "Calculate projected search direction and correction.")
+        self.optimization_utilities.CalculateRelaxedProjectedSearchDirectionAndCorrection(
+            nabla_f,
+            N,
+            omega_r,
+            omega_c,
+            solver,
+            p,
+            c)
+
+        # additional normalization step
+
+        self.optimization_utilities.AssignVectorToVariable(self.design_surface, p, KSO.PROJECTION)
+        self.p_norm = self.optimization_utilities.ComputeMaxNormOfNodalVariable(self.design_surface, KSO.PROJECTION)
+
+        # p *= 1.0 / self.p_norm
+        self.optimization_utilities.AssignVectorToVariable(self.design_surface, p+c, KSO.SEARCH_DIRECTION)
+        self.optimization_utilities.AssignVectorToVariable(self.design_surface, p+c, KSO.CONTROL_POINT_UPDATE)
```

Comment:
if the normalization does not happen you can remove these lines?

Reply:
it is true. I have revised the function.